File: hw2.py
### Econ 26610 Homework 2 ###

# Import packages
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from statsmodels.nonparametric.smoothers_lowess import lowess

# Load datasets
df_pre = pd.read_csv("/Users/jacksonvanvooren/Downloads/data/data_pre.csv")
df_post = pd.read_csv("/Users/jacksonvanvooren/Downloads/data/data_post.csv")
df_funding = pd.read_csv("/Users/jacksonvanvooren/Downloads/data/data_fundingcenters.csv")

#  Add policy period
df_pre["policyperiod"] = "pre"
df_post["policyperiod"] = "post"

# ---------------------------------------------------------------------------------------------- #
# 2

# Merge data for combined sample
df_merged = pd.concat([df_pre, df_post])

# Heating column
print(df_merged["heating"].value_counts(normalize=True))
# Mean age
print("Mean age:", df_merged["age"].mean())
# Mean land area
print("Mean land area:", df_merged["landarea"].mean())

# ---------------------------------------------------------------------------------------------- #
# 3
# Pre policy
# Create indicators
df_pre = pd.get_dummies(df_pre, columns=["year"], prefix="year")
df_pre["centralac"] = df_pre["centralac"].apply(lambda x: 1 if x == 1 else 0)
df_pre = pd.get_dummies(df_pre, columns=["exterior"], prefix="exterior")
df_pre = pd.get_dummies(df_pre, columns=["heating"], prefix="heating")
df_pre = pd.get_dummies(df_pre, columns=["condition"], prefix="condition")

X = df_pre[["year_1993", "year_1994", "year_1995", "year_1996",
            "year_1997", "centralac", "exterior_Brick", "exterior_Vinyl", 
            "heating_Central", "heating_Hot Water", "livingarea", "age", 
            "landarea", "condition_Average", "condition_Poor", "condition_Very Poor", 
            "bath"]]
y = df_pre["lpricesqft"]
reg_pre = LinearRegression().fit(X, y)
pre_coefficients = np.append(reg_pre.coef_, reg_pre.intercept_)
print(pre_coefficients)

# Post policy
# Create indicators
df_post = pd.get_dummies(df_post, columns=["year"], prefix="year")
df_post["centralac"] = df_post["centralac"].apply(lambda x: 1 if x == 1 else 0)
df_post = pd.get_dummies(df_post, columns=["exterior"], prefix="exterior")
df_post = pd.get_dummies(df_post, columns=["heating"], prefix="heating")
df_post = pd.get_dummies(df_post, columns=["condition"], prefix="condition")

X = df_post[["year_1999", "year_2000", "year_2001", "year_2002",
            "year_2003", "centralac", "exterior_Brick", "exterior_Vinyl", 
            "heating_Central", "heating_Hot Water", "livingarea", "age", 
            "landarea", "condition_Average", "condition_Poor", "condition_Very Poor", 
            "bath"]]
y = df_post["lpricesqft"]
reg_post = LinearRegression().fit(X, y)
post_coefficients = np.append(reg_post.coef_, reg_post.intercept_)
print(post_coefficients)

# ---------------------------------------------------------------------------------------------- #
# 5
# Cleaning data to only include funding center neighborhoods
df_merged_cleaned = df_merged.dropna(subset=['neighborhood'])

funding_neighborhoods = df_funding['neighborhood'].unique()
df_merged_cleaned = df_merged_cleaned[df_merged_cleaned['neighborhood'].isin(funding_neighborhoods)]

def calculate_euclidean(geo_x_1, geo_x_2, geo_y_1, geo_y_2):
    """Calulcates standard Euclidean distance between two locations"""
    return np.sqrt((geo_x_2 - geo_x_1)**2 + (geo_y_2 - geo_y_1)**2)

unique_neighborhoods = df_merged_cleaned['neighborhood'].unique()
all_distances = []

for neigh in unique_neighborhoods:
    df_homes_neigh = df_merged_cleaned[df_merged_cleaned['neighborhood'] == neigh]
    df_funding_neigh = df_funding[df_funding['neighborhood'] == neigh]
    
    neighborhood_distances = []
    for _, home_row in df_homes_neigh.iterrows():
        home_distances = []

        for _, funding_row in df_funding_neigh.iterrows():
            distance = calculate_euclidean(home_row["geo_x"], funding_row["geo_x"],
                                           home_row["geo_y"], funding_row["geo_y"])
            home_distances.append(distance)

        min_distance = min(home_distances)
        neighborhood_distances.append(min_distance)
    
    print(f"{neigh} average: {np.mean(neighborhood_distances)}")
    all_distances.extend(neighborhood_distances)

    # Plot the histogram
    bins = np.arange(min(neighborhood_distances), max(neighborhood_distances) + 50, 50)
    plt.hist(neighborhood_distances, bins=bins, edgecolor='black')
    plt.xlabel('Distance')
    plt.ylabel('Frequency')
    plt.title(f'Closest Distances for {neigh} (bin width 50)')
    plt.show()

df_merged_cleaned['min_distance'] = all_distances
print('Total average', np.mean(df_merged_cleaned['min_distance']))

# ---------------------------------------------------------------------------------------------- #
# 6
# Set up figure
plt.figure(figsize=(12, 10))
neighborhoods = df_merged_cleaned['neighborhood'].unique()

# LOWESS smoothing for each neighborhood by policy period
for i, neigh in enumerate(neighborhoods, start=1):
    df_neigh = df_merged_cleaned[df_merged_cleaned['neighborhood'] == neigh]
    ax = plt.subplot(2, 2, i)
    
    for period in ['pre', 'post']:
        df_period = df_neigh[df_neigh['policyperiod'] == period]
        smoothed = lowess(df_period['lpricesqft'], df_period['min_distance'])

        plt.plot(smoothed[:, 0], smoothed[:, 1], color='blue' if period == 'pre' else 'red',
                 label=f'{period.capitalize()}-Policy')

    plt.title(f'Lowess Smoothing: Pre vs Post Policy in {neigh}')
    plt.xlabel('Min Distance')
    plt.ylabel('Log Price per Sqft')
    plt.xlim(0, 1500)
    plt.legend()
    plt.grid(True)

File: hw2.R
# Econ 26610 Homework 2
# Jackson Van Vooren, Khwaish Vohra, Katherine Chen

# Load packages
library(stargazer)
library(dplyr)
library(ggplot2)
library(gridExtra)

# Load data
data_pre <- read.csv("/Users/jacksonvanvooren/Downloads/data/data_pre.csv")
data_post <- read.csv("/Users/jacksonvanvooren/Downloads/data/data_post.csv")
data_fundingcenters <- read.csv("/Users/jacksonvanvooren/Downloads/data/data_fundingcenters.csv")

#------------------------------------------------------------------------------#

# 2
# Assign categories (useful for LOESS smoothing later)
data_pre$policy_period <- "Pre-Policy"
data_post$policy_period <- "Post-Policy"

# Merge and analyze data
merged_data <- rbind(data_pre, data_post)
prop.table(table(merged_data$heating))
mean(merged_data$age)
mean(merged_data$landarea)

#------------------------------------------------------------------------------#

# 3
# Pre policy
# Assign base categories
df_pre <- data_pre %>%
  mutate(
    year = relevel(as.factor(year), ref = "1998"),
    exterior = relevel(as.factor(exterior), ref = "Wood or other"),
    heating = relevel(as.factor(heating), ref = "Gas"),
    condition = relevel(as.factor(condition), ref = "Good")
  )
df_pre <- df_pre %>%
  mutate(centralac = ifelse(centralac == 1, 1, 0))

# Extract dependend variable
y <- df_pre$lpricesqft  

# Dummies
X <- model.matrix(~ year + exterior + heating + condition + livingarea + age +
                    landarea + bath + centralac, data = df_pre) %>%
  as.data.frame()

# Regress
model_pre <- lm(y ~ . - 1, data = X)
summary(model_pre)


# Post policy
# Assign base categories
df_post <- data_post %>%
  mutate(
    year = relevel(as.factor(year), ref = "2004"),
    exterior = relevel(as.factor(exterior), ref = "Wood or other"),
    heating = relevel(as.factor(heating), ref = "Gas"),
    condition = relevel(as.factor(condition), ref = "Good")
  )
df_post <- df_post %>%
  mutate(centralac = ifelse(centralac == 1, 1, 0))

# Extract dependend variable
y <- df_post$lpricesqft  

# Dummies
X <- model.matrix(~ year + exterior + heating + condition + livingarea + age +
                    landarea + bath + centralac, data = df_post) %>%
  as.data.frame()

# Regress
model_post <- lm(y ~ . - 1, data = X)
summary(model_post)

# Make Latex tables
stargazer(model_pre, title="Pre-Policy Results", align=TRUE)
stargazer(model_post, title="Pre-Policy Results", align=TRUE)

#------------------------------------------------------------------------------#

# 5
# Initialize list and clean data to relevant neighborhoods
hist_plots <- list()
df_merged_cleaned <- merged_data %>%
  filter(!is.na(neighborhood)) %>%
  filter(neighborhood %in% data_fundingcenters$neighborhood)

# Helper to calculate Euclidean distance
calculate_euclidean <- function(x1, x2, y1, y2) {
  sqrt((x2 - x1)^2 + (y2 - y1)^2)
}

unique_neighborhoods <- unique(df_merged_cleaned$neighborhood)
min_distances <- numeric(nrow(df_merged_cleaned))

# Loop through neighborhoods with funding centers
for (neigh in unique_neighborhoods) {
  df_homes_neigh <- df_merged_cleaned %>% filter(neighborhood == neigh)
  df_funding_neigh <- data_fundingcenters %>% filter(neighborhood == neigh)

  neighborhood_distances <- numeric(nrow(df_homes_neigh))

  # Loop through homes in the corresponding neighborhood
  for (i in 1:nrow(df_homes_neigh)) {
    home_row <- df_homes_neigh[i, ]
    home_distances <- numeric(nrow(df_funding_neigh))
    
    # Find all distances to the funding center(s) for each home
    for (j in 1:nrow(df_funding_neigh)) {
      funding_row <- df_funding_neigh[j, ]
      distance <- calculate_euclidean(home_row$geo_x, funding_row$geo_x, 
                                      home_row$geo_y, funding_row$geo_y)
      home_distances[j] <- distance
    }
    
    # Take min distance for each home
    neighborhood_distances[i] <- min(home_distances)
  }
  
  # Average min distance for neighborhood
  cat(sprintf("%s average: %.2f\n", neigh, mean(neighborhood_distances)))
  
  # Update min_distances
  min_distances[df_merged_cleaned$neighborhood == neigh] <- neighborhood_distances
  
  # Plot histogram for each neighborhood
  p <- ggplot(data.frame(distance = neighborhood_distances), aes(x = distance)) +
    geom_histogram(binwidth = 50, color = "black", fill = "blue") +
    labs(title = paste("Minimum Distances for", neigh),
         x = "Distance", y = "Frequency") +
    theme_minimal()
  hist_plots[[neigh]] <- p
}

# Update df_merged_cleaned with the correct min_distance values
df_merged_cleaned$min_distance <- min_distances

# Total average
cat(sprintf("Total average: %.2f\n", mean(df_merged_cleaned$min_distance)))

# Plot histograms together in one figure
grid.arrange(grobs = hist_plots, ncol = 2)

#------------------------------------------------------------------------------#

# 6
# Plot using geom_smooth LOESS
p <- ggplot(df_merged_cleaned, aes(x = min_distance, y = lpricesqft,
                                   color = policy_period)) +
  geom_smooth(method = "loess", se = FALSE) +
  labs(title = "Log Land Price per Sqft vs. Distance to Funding Center",
       x = "Distance to Closest Funding Center",
       y = "Log Land Price per Sqft") +
  xlim(0, 1500) +
  theme_minimal() +
  scale_color_manual(values = c("Pre-Policy" = "blue", "Post-Policy" = "red")) +
  theme(legend.title = element_blank()) +
  facet_wrap(~neighborhood)

print(p)
